controllers/topology/reconcile_state.go

```diff
@@ -56,6 +58,69 @@ func (r *ClusterReconciler) reconcileInfrastructureCluster(ctx context.Context,
+	if ok, err := r.classRequiresInfrastructureMachineTemplate(ctx, desired); err == nil && ok {
+		cpInfraRef, err := getNestedRef(desired.controlPlane.object, "spec", "machineTemplate", "infrastructureRef")
+		if err != nil {
+			return errors.Wrapf(err, "failed to update the %s object,", desired.controlPlane.infrastructureMachineTemplate.GetKind())
+		}
+
+		// Create or update the MachineInfrastructureTemplate fo the control plane.
+		log.Info("updating", desired.controlPlane.infrastructureMachineTemplate.GroupVersionKind().String(), desired.controlPlane.infrastructureMachineTemplate.GetName())
+		cleanup, err = r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+			ref:                  cpInfraRef,
+			current:              current.controlPlane.infrastructureMachineTemplate,
+			desired:              desired.controlPlane.infrastructureMachineTemplate,
+			compatibilityChecker: checkReferencedObjectsAreCompatible,
+			// TODO: Use namer from PR 5070
+			templateNamer: func() string {
+				return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-infrastructure", desired.controlPlane.infrastructureMachineTemplate.GetName()))
+			},
+		},
+		)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update the %s object", desired.controlPlane.infrastructureMachineTemplate.GetKind())
+		}
+	}
+	// Create or update a ControlPlaneObject for the controlPlaneTopologyState.
+	log.Info("updating", desired.controlPlane.object.GroupVersionKind().String(), desired.controlPlane.object.GetName())
+	if err := r.reconcileReferencedObject(ctx, current.controlPlane.object, desired.controlPlane.object); err != nil {
+		return kerrors.NewAggregate([]error{errors.Wrapf(err, "failed to update the %s object", desired.controlPlane.object.GetKind()), cleanup()})
+	}
+
+	// At this point we've updated the ControlPlane object and, where required, the ControlPlane InfrastructureMachineTemplate
+	// without error. Run the cleanup in order to delete the old InfrastructureMachineTemplate if template rotation was done during update.
+	return cleanup()
+}
+
+// classRequiresInfrastructureMachineTemplate assesses whether or not a given cluster requires an InfrastructureMachineTemplate as part of the Cluster ControlPlane
+// TODO: Make this method used more widely across the code to check for this specific condition. There's a possible issue with using get here - getControlPlaneState uses a class passed to it.
+func (r *ClusterReconciler) classRequiresInfrastructureMachineTemplate(_ context.Context, desired *clusterTopologyState) (bool, error) {
+	if 2 == 1 {
+		return false, errors.New("mock error for linting")
+	}
+	// This is the basic check to see if an InfrastructureMachineTemplate reconciliation is needed.
+	return desired.controlPlane.infrastructureMachineTemplate != nil, nil
+
+	// TODO: Find out if it's worth the additional complexity to rewrite the check in this way.
```

Comment:
This is something I'd like pointers on - the check as is - using only the existence of the Infrastructure in the desired state - may not be sufficient to cover edge cases, but checking the single source of truth - the clusterClass - for this value either means passing the class down the code or else running a client.Get in the middle of the reconcile function which complicates the error states and testing significantly.

Reply:
What edge cases do we have?

I would have assumed it's the responsibility of the `computeControlPlane` part to set `infrastructureMachineTemplate` correctly so we don't have to double check in reconcile. 

I think we shouldn't call `getClass` again in any case. In my opinion there are just too many things which could have changed in the mean time to handle that correctly. I think it gets really really complicated if part of our reconcile code is based on a later version of our resources.

Reply:
+1 to what @sbueringer said. 
Given that we are already using `class.clusterClass.Spec.ControlPlane.MachineInfrastructure != nil` everywhere as a check to see if we need an infra template in the control plane and the documentation for the field also suggests 
https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha4/clusterclass_types.go#L63-L70
the same thing I think it is fine to have a simple check instead of getting the cluster class again from the api server (could potentially bring an updated version of the cluster class).


> I think it gets really really complicated if part of our reconcile code is potentially based on a later version of relevant resources.

+1

Reply:
What about having

```
func (c *controlPlaneTopologyClass) HasInfrastructureMachine() bool {
	return c.infrastructureMachineTemplate != nil
}
```
So our code can be changed into 

```
func foo(class *clusterTopologyClass)  {
	if class.controlPlane.HasInfrastructureMachine() {
		// do something
	}
}
```

Reply:
I've updated the way this was done to bring the controlPlaneTopologyClass down the call chain and implemented a check like this. I'm still not really happy with the solution so it definitely bears further discussion.